Stackwalk support for interpreted frames

janvorli opened the pull request:
This change adds support for walking stack with interpreter frames present. The `StackFrameIterator` now considers the interpreter frames to be frameless frames like regular JITted / crossgened folder. This is a basis for GC, EH and debugger stack walking. The SOS `clrstack` and `clrstack -i` both display correct stack trace with interpreter frames on the stack with this change.

It required changing the `GetCallerSP` and `EnsureCallerContextIsValid` to become virtual methods so that the right variant for the interpreter or JITted / AOTed code can be invoked. Before, they were both static methods on the `EECodeManager`.

I have also added some extra handling of the case when the interpreter frames are on top of the stack so that the clrstack can dump them.

janvorli committed:
Stackwalk support for interpreted frames

This change adds support for walking stack with interpreter frames
present. The StackFrameIterator now considers the interpreter frames to
be frameless frames like regular JITted / crossgened folder.
This is a basis for GC, EH and debugger stack walking. The SOS
`clrstack` and `clrstack -i` both display correct stack trace with
interpreter frames on the stack with this change.

I have added some extra handling of the case when the interpreter frames
are on top of the stack so that the clrstack can dump them.

janvorli commented:
cc: @BrzVlad, @cshung

kg commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -1841,7 +1841,7 @@ CLRUnwindStatus ExceptionTracker::ProcessOSExceptionNotification(
-    uCallerSP = EECodeManager::GetCallerSp(cfThisFrame.pRD);
+    uCallerSP = cfThisFrame.GetCodeManager()->GetCallerSp(cfThisFrame.pRD);
 #else // !ESTABLISHER_FRAME_ADDRESS_IS_CALLER_SP
     uCallerSP = sf.SP;
```
Do we need to do something in this other arm (looks like it's for ARM and ARM64 specifically) to support interp?

janvorli replied:
This is the legacy EH code, this is not going to work with the interpreter anyways. Actually looking at the GetCallerSP usage, it seems that don't have a place where we would actually call it for the interpreter frame. I have made the change of `GetCallerSP` and `EnsureCallerContextIsValid` together, but maybe I could revert the `GetCallerSP`

jkotas replied:
> maybe I could revert the GetCallerSP

Do you plan to do that?

janvorli replied:
@kotas done in the last commit

jkotas commented on src/coreclr/debug/daccess/dacdbiimplstackwalk.cpp:
```diff
@@ -185,7 +189,23 @@ void DacDbiInterfaceImpl::SetStackWalkCurrentContext(VMPTR_Thread           vmTh
+        InterpreterEntryFrame *pEntryFrame = dac_cast<PTR_InterpreterEntryFrame>(pTopFrame);
+        PTR_InterpMethodContextFrame pTOSInterpMethodContextFrame = pEntryFrame->GetInterpMethodTopmostContextFrame();
+        SetIP(pContext2, (TADDR)pTOSInterpMethodContextFrame->ip);
+        SetSP(pContext2, dac_cast<TADDR>(pTOSInterpMethodContextFrame));
```
Do you think that overloading SP to mean interpreter stack pointer when we are executing interpreter code is going to work well?

For example, it means that we won't be able to simple compare two stack pointers to determine relative position of two frames on the stack. We will have to walk the whole chain instead.

janvorli replied:
Why? The SPs are still ordered correctly.

jkotas replied:
Ah ok, I got confused by the reuse scheme.

janvorli replied:
To explain in more detail, the SP is actually not pointing to the interpreter stack in the sense of the stack that interpreter uses for its locals and call arguments. It is a pointer to the `InterpMethodContextFrame` structures that are allocated on regular stack using `alloca` at each interpreter call (they are reused when the interpreter call returns and another one happens).

jkotas commented on src/coreclr/vm/FrameTypes.h:
```diff
@@ -47,6 +47,10 @@ FRAME_TYPE_NAME(DebuggerClassInitMarkFrame)
 FRAME_TYPE_NAME(DebuggerU2MCatchHandlerFrame)
 FRAME_TYPE_NAME(ExceptionFilterFrame)
+#ifdef FEATURE_INTERPRETER
+FRAME_TYPE_NAME(InterpreterEntryFrame)
```
Do we need really to have both entry and exit frames? I would expect that one `InterpreterFrame` should be enough. We will know that we have exited the interpreter if we find some other Frame or managed code first on the stack.

janvorli replied:
I was considering that and this seems to work better with the stack walker. We could possibly have just the `InterpreterEntryFrame`, but then upon reaching it, we would immediately need to switch to the underlying interpreter frames and then once we get out of them, hit the `InterpreterEntryFrame` again and copy the transition record to the context to move to the caller of the sequence of interpreted frames. Also, the `InterpreterEntryFrame` doesn't have a direct pointer to the topmost frame in the `InterpExecMethod`, Vlad was concerned about updating that on every interpreter to interpreter call.

jkotas replied:
Are the interpreter -> AOT code calls going to push InterpreterExitFrame?

I think we should be optimizing for making the interpreter -> AOT calls as cheap as possible.

janvorli replied:
Yes, those would need it too. We can pass current thread as a parameter to the InterpMethodFrame and then use it to push that frame without the thread local access overhead.

BrzVlad replied:
So the purpose of the `InterpreterEntryFrame` is to store the context so can unwind to AOT callers, while the purpose of the `InterpreterExitFrame` is to make it cleaner to access the top most interpreter frame ? Would we always resume execution in the interpreter loop via a try/catch ? So if EH determines that an interpreter method caught the exception, in order to resume in the interpreter it would throw an exception that will get caught at this interpreter exit location ?

jkotas commented on src/coreclr/vm/frames.cpp:
```diff
@@ -2079,6 +2083,40 @@ PCODE UnmanagedToManagedFrame::GetReturnAddress_Impl()
+    PTR_InterpMethodContextFrame pFrame = m_pInterpMethodContextFrame;
+    _ASSERTE(pFrame != NULL && pFrame->ip != NULL);
+
+    // The frames of a method are linked in a reverse order (from bottom to top of the part of the stack)
```
Is there an advantage in this reverse order? I would expect the interpreter frame to point to what we are executing.

janvorli replied:
Vlad was concerned about updating such a pointer on every interpreter to interpreter call. The current call implementation is pretty efficient and according to his experience from Mono, this would add noticeable overhead.

janvorli replied:
Moreover, this reverse walk is only needed in the native debugger on desktop scenario when someone has the debugger sitting in the middle of the `InterpExecMethod`, nothing else needs to walk that way (provided we have the two interpreter frame kinds).

jkotas replied:
> Vlad was concerned about updating such a pointer on every interpreter to interpreter call.

We do not need to update the pointer on every interpreter-to-interpreter call. We only need to update the pointer when we exit the interpreter. ie instead of pushing the full Frame to exit the interpreter, we would just need to patch the current IP in the Interpreter frame.

janvorli replied:
I think that since only the native debugger needs to walk this list and it would need to walk it even if I changed it the way you suggested in case it was sitting in the `InterpExecMethod` and someone has invoked the `clrstack` command, it doesn't seem really beneficial to store it. Unless we moved to the single interpreter frame way, which would introduce the unusual double handling of single explicit frame to the `StackFrameIterator` or something similar to that.
What is your concern about having two separate explicit frames?

jkotas replied:
I think the scheme looks more expensive than it needs to be and non-intuitive (the linked list is in wrong direction).

janvorli replied:
Yes, the direction I am using for this specific native debugger scenario is primarily used for the interpreter internal purposes, I have just reused it for these rare scenarios. This is not used for managed debuggers, GC or EH. In those cases we would always have the interpreter exit frame.

BrzVlad commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -2750,10 +2750,16 @@ extern "C" void STDCALL ExecuteInterpretedMethod(TransitionBlock* pTransitionBlo
 
+    InterpreterEntryFrame interpreterEntryFrame(pTransitionBlock, &interpFrame);
+
+    // Interpreter-TODO: catch exceptions stemming from the InterpExecMethod and call the managed exception handling
```
Not sure what you mean here by catching exceptions thrown by `InterpExecMethod`. No C++ exceptions should ever be thrown at this point.

janvorli replied:
Not that method itself, but its callees can throw.

BrzVlad replied:
But `InterpExecMethod` should catch all C++ exceptions and convert them to managed exceptions if necessary, right ? If we reach here, we got rid of all managed frames and it would be incorrect. Just to clarify.

janvorli committed:
Fix VirtualUnwindInterpreterCallFrame

In case we were generating the caller context, the SP and ContextFlags
were missing.

janvorli committed:
Move to single InterpreterFrame and few fixes

janvorli committed:
Fix Unix build break

janvorli committed:
Prevent missing override errors on Unix build

janvorli committed:
Improve top frame lookup

janvorli commented:
@jkotas I have made changes based on your feedback. There is now a single `InterpreterFrame` and it keeps pointer to the top frame of the related `InterpExecMethod`. That pointer is updated whenever the `InterpExecMethod` calls something external where runtime can be suspended. To enable it to work with SOS when debugging dumps or live debugging and the interpreter running in the `InterpExecMethod` without leaving it, we still search for the real top starting at the value that's stored there. In the EH, GC and managed debugger cases, the pointer will always be the right one.

janvorli commented:
The CI test failures are unrelated.

jkotas commented on src/coreclr/inc/eetwain.h:
```diff
@@ -396,7 +403,6 @@ TADDR GetAmbientSP(PREGDISPLAY     pContext,
     Get the number of bytes used for stack parameters.
     This is currently only used on x86.
 */
-virtual
```
Is there a pateren for why `virtual` was removed on some methods and not on the others?

For example, why `GetAmbientSP` a few lines above has virtual, but `GetStackParameterSize` lost it?

janvorli replied:
That was actually not intentional. The way it happened was that I've first converted all methods to use "override" instead of just "virtual", but the compilation on Linux started to fail because the compiler when it saw override somewhere wanted it everywhere and not just on the classes I had it on, but it wanted it on the base class and I believe also on some related classes. So I've reverted that to not to expand the change by unnecessary changes, but missed couple of places were I've removed the virtual.

jkotas commented on src/coreclr/vm/gcinfodecoder.cpp:
```diff
@@ -19,7 +19,11 @@
 #ifndef GET_CALLER_SP
-#define GET_CALLER_SP(pREGDISPLAY) EECodeManager::GetCallerSp(pREGDISPLAY)
+inline size_t GET_CALLER_SP(PREGDISPLAY pREGDISPLAY)
+{
```
Is this method actually used? It looks like a left-over from JIT64 era. Can we turn it into `ASSERT(false); return NULL;` or something like that?

janvorli replied:
I've added the ASSERT here and it hasn't fired while running coreclr / libraries tests, so it looks like it is really obsolete. I'll commit that change.

jkotas commented on src/coreclr/vm/gcinfodecoder.cpp:
```diff
@@ -2140,6 +2144,9 @@ template <typename GcInfoEncoding> OBJECTREF* TGcInfoDecoder<GcInfoEncoding>::Ge
     }
     else if( GC_CALLER_SP_REL == spBase )
     {
+#ifndef FEATURE_NATIVEAOT
+        _ASSERTE(pRD->IsCallerContextValid);
+#endif
```
```suggestion
```
`GET_CALLER_SP` does this validation, so we do not need to repeat it here. Also, I suspect that this would be a build break for #113888.

janvorli replied:
I have added this validation before adding that function and somehow missed this was the only place where the `GET_CALLER_SP` was used. I'll remove it.

jkotas commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -2746,12 +2746,16 @@ extern "C" void STDCALL ExecuteInterpretedMethod(TransitionBlock* pTransitionBlo
-    InterpExecMethod(&interpFrame, threadContext);
+    InterpreterFrame interpreterFrame(pTransitionBlock, &interpMethodContextFrame);
+
+    InterpExecMethod(&interpreterFrame, &interpMethodContextFrame, threadContext);
```
This change looks ok for now. Do we have a plan for where and how the arguments are going to converted from the `TransitionBlock` to the interpreter representation?

BrzVlad replied:
I'm thinking when invoking an interpreted method through the precode thunks, instead of calling `ExecuteInterpretedMethod` like we do now, it will call some other assembly thunk that receives a lower level bytecode referenced from the interpreter bytecode (maybe in some Interpreter code header). These assembly thunks will move arguments from the transition block to the interpreter stack, finally dispatching to `ExecuteInterpreterMethod`. I described this mechanism a while ago in a doc. Last time I spoke with @janvorli, this plan was to use this mechanism from the start since we will need it as a fallback anyway on iOS.

janvorli replied:
I've used the TransitionBlock here for two reasons. One, as a possibly easy starting point for passing arguments until we implement the proper mechanism and second, as a mean to save and restore the callee saved registers. This is necessary as we need to be able to restore them during stack walking from interpreted to JITted / AOTed methods. Once we have the optimized mechanism, we can use a smaller structure than the TransitionBlock that would contain only the callee saved registers.

janvorli committed:
Reflect PR feedback

* Change the GET_CALLER_SP to assert and return NULL
* Remove extra assert at the call to GET_CALLER_SP
* Revert some changes removing "virtual" keyword that were not
  intentional

janvorli committed:
Few missing virtual keywords reverts

janvorli committed:
Revert the GetCallerSP changes

janvorli committed:
Fix MUSL build break

jkotas commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -698,6 +706,7 @@ void InterpExecMethod(InterpMethodContextFrame *pFrame, InterpThreadContext *pTh
                     MethodDesc *pMD = (MethodDesc*)(targetMethod & ~INTERP_METHOD_DESC_TAG);
                     PCODE code = pMD->GetNativeCode();
                     if (!code) {
+                        pInterpreterFrame->SetTopInterpMethodContextFrame(pFrame);
```
If I understand this correctly, this is just an optimization and not required for correctness. Is that correct? It may be worth mentioning it in a comment.

Does the setting need to be undone when the call returns?

It is not clear to me whether this optimization is worth it. It may be better to do it during stackwalk - so that repeated stackwalks from approximately the same spot avoid walking the whole list.

janvorli replied:
Yes, it would still work correctly without it, so it can be viewed as an optimization.

It doesn't need to be undone when the call returns:
* for the SOS scenario, we would seek for the top frame going in one way or the other through the linked list from the last value stored there. It doesn't matter if we start on a frame that's in the middle of the "substack" belonging to the current InterpExecMethod or if we start on an inactive frame that's somewhere by the end of the list. Based on the inactive vs active state of the frame (ip being 0 or non-zero), we know which way to seek for the last active frame.
* for other scenarios, we would maintain the exact value so that we don't waste time scanning the list.

Since this optimization is just one store, I think it is worth doing that instead of updating it during the stack walk. Imagine a case with a deep recursion within interpreted frames only. It would cause a lot of seeking the first time we do a stack walk which is completely unnecessary.
We will need to do other stuff while leaving the interpreter, like catching exceptions from the native code and calling the DispatchManagedException, so this is going to be just one of the things to do. I am planning to add a macro named "EXIT_INTERPRETER_BEGIN/END or something along those lines to contain that stuff.

jkotas replied:
> It would cause a lot of seeking the first time we do a stack walk which is completely unnecessary.

This work is small and proportional to the total cost of stack walk (assuming we typically walk large part of the stack). 

I am looking at it as "total number of instructions executed" in given scenario. I think updating the pointer in the stackwalker would be better on average in this metric.

This is a micro-optimization. It is fine to keep this as is. Comment would be nice.

> We will need to do other stuff while leaving the interpreter, like catching exceptions from the native code and calling the DispatchManagedException, so this is going to be just one of the things to do. 

Yep. I expect that the total cost of this other stuff will make us consider building the interpreter executor in C# eventually - but that is for future discussion.

janvorli replied:
I've added a comment

janvorli committed:
Add comment on top frame seeking

kg commented on src/coreclr/vm/eetwain.cpp:
```diff
@@ -2277,13 +2281,78 @@ ULONG32 EECodeManager::GetStackParameterSize(EECodeInfo * pCodeInfo)
+    {
+        // We already have the caller's frame context
+        // We just switch the pointers
+        PT_CONTEXT temp      = pRD->pCurrentContext;
```
Isn't unwinding the process of walking up the stack? Why are we switching the pointers instead of just overwriting the current context with the caller context? I assume I'm missing something here

janvorli replied:
If we already have the caller context available, then you can just set the current context to it, there is no need to call unwind again to get the same values. The switching of pointer is to avoid copying in this specific case. 
There are cases when we want to know the caller context without actually unwinding, so we call EnsureCallerContextIsValid. That's where the caller context gets created and then it can get reused here.

kg reviewed (approved):
LGTM

janvorli commented:
/ba-g the test failures in this PR are happening on all PRs.
